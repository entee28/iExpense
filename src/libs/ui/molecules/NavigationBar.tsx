import { faChevronLeft } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'
import { useNavigation } from '@react-navigation/native'
import { NAVBAR_HEIGHT } from 'libs/ui/constants'
import { getWidth } from 'libs/utils'
import React from 'react'
import { StyleProp, StyleSheet, View, ViewStyle } from 'react-native'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import { Pressable, Text } from '../atoms'
import colors from '../colors'

type Props = {
  left?: JSX.Element | null
  right?: JSX.Element
  title?: string | JSX.Element
  shadow?: boolean
  leftStyle?: StyleProp<ViewStyle>
  rightStyle?: StyleProp<ViewStyle>
  transparent?: boolean
  backgroundColor?: string
  color?: string
  onBack?: () => void
  border?: boolean
  style?: StyleProp<ViewStyle>
  testIDScreen?: string
}

const SIDE_WIDTH = 90

export const NavigationBar = ({
  left,
  right,
  title,
  leftStyle,
  rightStyle,
  shadow = false,
  onBack,
  backgroundColor = colors.white,
  transparent = false,
  color = colors.mono100,
  testIDScreen = '',
  border,
  style
}: Props) => {
  const { top } = useSafeAreaInsets()

  const navigation = useNavigation()

  const goBack = () => {
    onBack ? onBack() : navigation.goBack()
  }

  return (
    <View
      testID={`navigation_bar.${testIDScreen}`}
      style={[
        styles.root,
        shadow && styles.shadow,
        border && styles.border,
        {
          paddingTop: top,
          height: NAVBAR_HEIGHT + top,
          backgroundColor: transparent ? 'transparent' : backgroundColor
        },
        style
      ]}>
      <View style={[styles.left, leftStyle]}>
        {left !== undefined ? (
          left
        ) : (
          <Pressable
            testID="navigation_bar.left"
            onPress={goBack}
            style={styles.backButton}>
            <FontAwesomeIcon icon={faChevronLeft} size={14} color={color} />
          </Pressable>
        )}
      </View>

      <View style={styles.center}>
        {typeof title === 'string' ? (
          <Text
            testID={`navigation_bar.${testIDScreen}_title`}
            bold
            fontSize={16}
            lineHeight={24}
            textAlign="center"
            color={color}>
            {title}
          </Text>
        ) : (
          title
        )}
      </View>

      <View style={[styles.right, rightStyle]}>{right}</View>
    </View>
  )
}

const styles = StyleSheet.create({
  root: {
    width: '100%',
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: colors.white
  },
  left: {
    width: SIDE_WIDTH,
    alignItems: 'flex-start',
    paddingLeft: 24,
    height: '100%',
    justifyContent: 'center'
  },
  right: {
    width: SIDE_WIDTH,
    alignItems: 'flex-end',
    paddingRight: 24,
    height: '100%',
    justifyContent: 'center'
  },
  center: {
    width: getWidth(100) - 2 * SIDE_WIDTH,
    height: '100%',
    justifyContent: 'center'
  },
  shadow: {
    shadowOffset: { width: 0, height: 2 },
    shadowColor: colors.primary100,
    shadowOpacity: 0.1,
    shadowRadius: 4
  },
  backButton: {
    paddingHorizontal: 6,
    paddingVertical: 4
  },
  border: {
    borderBottomWidth: 1,
    borderBottomColor: colors.mono10
  }
})
