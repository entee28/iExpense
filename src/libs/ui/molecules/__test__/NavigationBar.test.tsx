import { faGear } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'
import { NavigationContainer } from '@react-navigation/native'
import { render } from '@testing-library/react-native'
import { NavigationBar } from '../NavigationBar'

it('should match snapshot', () => {
  const navbar = render(
    <NavigationBar
      title="Hello Test"
      right={<FontAwesomeIcon icon={faGear} />}
    />
  ).toJSON()

  expect(navbar).toMatchSnapshot()
})
