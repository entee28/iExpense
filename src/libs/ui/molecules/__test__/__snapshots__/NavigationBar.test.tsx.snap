// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "flexDirection": "row",
        "width": "100%",
      },
      false,
      undefined,
      {
        "backgroundColor": "#FFFFFF",
        "height": 44,
        "paddingTop": 0,
      },
      undefined,
    ]
  }
  testID="navigation_bar."
>
  <View
    style={
      [
        {
          "alignItems": "flex-start",
          "height": "100%",
          "justifyContent": "center",
          "paddingLeft": 24,
          "width": 90,
        },
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "paddingHorizontal": 6,
          "paddingVertical": 4,
        }
      }
    >
      <
        color="#1A1A1A"
        icon={
          {
            "icon": [
              384,
              512,
              [
                9001,
              ],
              "f053",
              "M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 278.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z",
            ],
            "iconName": "chevron-left",
            "prefix": "fas",
          }
        }
        size={14}
      />
    </View>
  </View>
  <View
    style={
      {
        "height": "100%",
        "justifyContent": "center",
        "width": 570,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#1A1A1A",
            "fontFamily": "Nunito-Bold",
            "fontSize": 16,
            "lineHeight": 24,
            "textAlign": "center",
          },
          {},
        ]
      }
      testID="navigation_bar._title"
    >
      Hello Test
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "flex-end",
          "height": "100%",
          "justifyContent": "center",
          "paddingRight": 24,
          "width": 90,
        },
        undefined,
      ]
    }
  >
    <
      icon={
        {
          "icon": [
            512,
            512,
            [
              9881,
              "cog",
            ],
            "f013",
            "M495.9 166.6c3.2 8.7 .5 18.4-6.4 24.6l-43.3 39.4c1.1 8.3 1.7 16.8 1.7 25.4s-.6 17.1-1.7 25.4l43.3 39.4c6.9 6.2 9.6 15.9 6.4 24.6c-4.4 11.9-9.7 23.3-15.8 34.3l-4.7 8.1c-6.6 11-14 21.4-22.1 31.2c-5.9 7.2-15.7 9.6-24.5 6.8l-55.7-17.7c-13.4 10.3-28.2 18.9-44 25.4l-12.5 57.1c-2 9.1-9 16.3-18.2 17.8c-13.8 2.3-28 3.5-42.5 3.5s-28.7-1.2-42.5-3.5c-9.2-1.5-16.2-8.7-18.2-17.8l-12.5-57.1c-15.8-6.5-30.6-15.1-44-25.4L83.1 425.9c-8.8 2.8-18.6 .3-24.5-6.8c-8.1-9.8-15.5-20.2-22.1-31.2l-4.7-8.1c-6.1-11-11.4-22.4-15.8-34.3c-3.2-8.7-.5-18.4 6.4-24.6l43.3-39.4C64.6 273.1 64 264.6 64 256s.6-17.1 1.7-25.4L22.4 191.2c-6.9-6.2-9.6-15.9-6.4-24.6c4.4-11.9 9.7-23.3 15.8-34.3l4.7-8.1c6.6-11 14-21.4 22.1-31.2c5.9-7.2 15.7-9.6 24.5-6.8l55.7 17.7c13.4-10.3 28.2-18.9 44-25.4l12.5-57.1c2-9.1 9-16.3 18.2-17.8C227.3 1.2 241.5 0 256 0s28.7 1.2 42.5 3.5c9.2 1.5 16.2 8.7 18.2 17.8l12.5 57.1c15.8 6.5 30.6 15.1 44 25.4l55.7-17.7c8.8-2.8 18.6-.3 24.5 6.8c8.1 9.8 15.5 20.2 22.1 31.2l4.7 8.1c6.1 11 11.4 22.4 15.8 34.3zM256 336c44.2 0 80-35.8 80-80s-35.8-80-80-80s-80 35.8-80 80s35.8 80 80 80z",
          ],
          "iconName": "gear",
          "prefix": "fas",
        }
      }
    />
  </View>
</View>
`;
